From 81a63d264340d0ae57fdc1d92cedf2db1e951509 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jo=C3=A3o=20Henrique=20Ferreira=20de=20Freitas?=
 <joaohf@gmail.com>
Date: Thu, 12 Sep 2019 22:05:22 +0200
Subject: [PATCH] Add support to cross compile using rebar3

 * remove rebar3_run plugin
 * remove observer load dependency
---
 rebar.config        |  5 ++--
 rebar.config.script | 59 +++++++++++++++++++++++++++++++++++++++------
 2 files changed, 55 insertions(+), 9 deletions(-)

diff --git a/rebar.config b/rebar.config
index 7b49f83..376724f 100644
--- a/rebar.config
+++ b/rebar.config
@@ -39,6 +39,8 @@
 
 {relx,
  [{include_src, false},
+  {include_erts, true},
+  {system_libs, true},
   {extended_start_script, false},
   {generate_start_script, false},
   {sys_config, false},
@@ -130,7 +132,6 @@
   {emqx_lua_hook, load},
   {emqx_lwm2m, load},
   {emqx_statsd, load},
-  {observer, load},
   luerl,
   xmerl
  ]}.
@@ -153,7 +154,7 @@
 {cover_opts, [verbose]}.
 {cover_export_enabled, true}.
 
-{plugins, [rebar3_run]}.
+{plugins, []}.
 
 {post_hooks,
  [{"(linux|darwin|solaris|freebsd|netbsd|openbsd)", compile, "./post-compile.sh"},
diff --git a/rebar.config.script b/rebar.config.script
index 69c1052..da5d80b 100644
--- a/rebar.config.script
+++ b/rebar.config.script
@@ -108,23 +108,68 @@ DefaultDepRef =
 %% Relx configs
 %% ==============================================================================
 
+ReplaceRelx =
+fun(Key, Path, Config) ->
+    case lists:keyfind(relx, 1, Config) of
+        false ->
+            Config;
+        {relx, Relx} ->
+            case lists:keyfind(Key, 1, Relx) of
+                false ->
+                    Config;
+                {Key, true} ->
+                    NewPath = {Key, Path},
+                    NewRelx = lists:keyreplace(Key, 1, Relx, NewPath),
+                    lists:keystore(relx, 1, Config, {relx, NewRelx});
+                {Key, _} ->
+                    Config
+            end
+    end
+end,
+
+ReplaceIncludeErts =
+fun(Path, Config) ->
+    ReplaceRelx(include_erts, Path, Config)
+end,
+
+ReplaceSystemLibs =
+fun(Path, Config) ->
+    ReplaceRelx(system_libs, Path, Config)
+end,
+
+EnvTargetIncludeErts = os:getenv("REBAR3_TARGET_INCLUDE_ERTS"),
+EnvTargetSystemLibs = os:getenv("REBAR3_TARGET_SYSTEM_LIBS"),
+
+Check = {EnvTargetIncludeErts, EnvTargetSystemLibs},
+CONFIGX = case Check of
+    {false, false} ->
+        CONFIG2;
+    {IncludeErts, false} ->
+        ReplaceIncludeErts(IncludeErts, CONFIG2);
+    {false, SystemLibs} ->
+        ReplaceSystemLibs(SystemLibs, CONFIG2);
+    {IncludeErts, SystemLibs} ->
+        _CONFIG = ReplaceIncludeErts(IncludeErts, CONFIG2),
+        ReplaceSystemLibs(SystemLibs, _CONFIG)
+end,
+
 GitDescribe = case DefaultDepRef of
                   {tag, EnvTag} -> EnvTag;
                   _Else -> lists:last(string:tokens(os:cmd("git tag -l \"v*\""), "\n"))
               end,
-Relx0 = Kf(relx, CONFIG2),
+Relx0 = Kf(relx, CONFIGX),
 {release, {_, Vsn0}, RelxBaseApps0} = lists:keyfind(release, 1, Relx0),
 Vsn1 = case Vsn0 of
     git_describe -> GitDescribe;
     Vsn -> Vsn
 end,
-RelxElixirApps = Kf(elixir_relx_apps, CONFIG2),
+RelxElixirApps = Kf(elixir_relx_apps, CONFIGX),
 RelxBaseApps = RelxBaseApps0 ++ RelxElixirApps,
 RelxOverlay = Kf(overlay, Relx0),
-RelxCloudApps = RelxBaseApps ++ Kf(cloud_relx_apps, CONFIG2),
-RelxEdgeApps = RelxBaseApps ++ Kf(edge_relx_apps, CONFIG2),
-RelxCloudOverlay0 = Kf(cloud_relx_overlay, CONFIG2),
-RelxEdgeOverlay0 = Kf(edge_relx_overlay, CONFIG2),
+RelxCloudApps = RelxBaseApps ++ Kf(cloud_relx_apps, CONFIGX),
+RelxEdgeApps = RelxBaseApps ++ Kf(edge_relx_apps, CONFIGX),
+RelxCloudOverlay0 = Kf(cloud_relx_overlay, CONFIGX),
+RelxEdgeOverlay0 = Kf(edge_relx_overlay, CONFIGX),
 RelxCloudOverlay = RelxOverlay ++ RelxCloudOverlay0,
 RelxEdgeOverlay = RelxOverlay ++ RelxEdgeOverlay0,
 
@@ -176,7 +221,7 @@ Deletes = [ deps
 
 Additions = [{profiles, Profiles}],
 
-CONFIG3 = lists:foldl(fun(K, Acc) -> lists:keydelete(K, 1, Acc) end, CONFIG2, Deletes),
+CONFIG3 = lists:foldl(fun(K, Acc) -> lists:keydelete(K, 1, Acc) end, CONFIGX, Deletes),
 
 CONFIG4 = lists:foldl(fun({K, V}, Acc) -> lists:keystore(K, 1, Acc, {K, V}) end, CONFIG3, Additions),
 
-- 
2.17.1

